File: rust/2. Files R/src/main.rs
use std::time::Instant;

use tokio::task::JoinHandle;
use tokio::fs;

async fn routine(file_index: i32) -> String {
    let path = format!("./data/{file_index}.txt");
    fs::read_to_string(path).await.unwrap()
}

fn lcg(n: usize, max_value: i128) -> Vec<i32> {
    let mut list: Vec<i128> = Vec::with_capacity(n);
    list.push(42);

    let m = 0x7FFFFFFF;
    let a = 48271;
    let c = 0;
    for i in 1..n {
        list.push((list[i - 1] * a + c) % m);
        list[i - 1] = list[i - 1] % max_value;
    }
    list[n - 1] = list[n - 1] % max_value;

    list.iter().map(|it| *it as i32).collect()
}

#[tokio::main]
async fn main() {
    tokio::spawn(routine(1)).await.unwrap();

    let n = std::env::args().nth(1).unwrap().parse().unwrap();
    let generated = lcg(n, 20);
    let indices = generated.iter();

    let start = Instant::now();

    let handles: Vec<JoinHandle<_>> = indices.map(|i| {
        tokio::spawn(routine(*i))
    }).collect();

    for handle in handles {
        handle.await.unwrap();
    }

    println!("{:?}", start.elapsed().as_micros());
}

File: go/2. Files R/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func routine(fileIndex int, ch chan<- string) {
	path := fmt.Sprintf("./data/%d.txt", fileIndex)
	file, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}

	ch <- string(file)
}

func lcg(n, maxValue int) []int {
	timers := make([]int, n)
	timers[0] = 42

	m := 0x7FFFFFFF
	a := 48271
	c := 0
	for i := 1; i < n; i++ {
		timers[i] = (timers[i-1]*a + c) % m
		timers[i-1] = timers[i-1] % maxValue
	}
	timers[n-1] = timers[n-1] % maxValue

	return timers
}

func main() {
	n, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic(err)
	}

	indices := lcg(n, 20)
	ch := make(chan string)

	start := time.Now()
	for i := 0; i < n; i++ {
		go routine(indices[i], ch)
	}

	list := make([]string, n)
	for i := 0; i < n; i++ {
		list[i] = <-ch
	}

	_ = len(list)

	fmt.Println(time.Since(start).Microseconds())
}
